Signal/src/util/TextFieldHelper.swift

```diff
@@ -0,0 +1,56 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+@objc class TextFieldHelper: NSObject {
+
+    // Used to implement the UITextFieldDelegate method: `textField:shouldChangeCharactersInRange:replacementString` but takes advantage of
+    // Swifts superior unicode handling to append partial pasted text without splitting multi-byte characters.
+    class func textField(_ textField: UITextField, shouldChangeCharactersInRange editingRange: NSRange, replacementString: String, byteLimit: UInt) -> Bool {
+
+        let byteLength = { (string: String) -> UInt in
+            return UInt(string.utf8.count)
+        }
+
+        let existingString = textField.text ?? ""
+
+        // Given an NSRange, we need to interact with the NS flavor of substring
+        let removedString = (existingString as NSString).substring(with: editingRange)
+
+        let lengthOfRemainingExistingString = byteLength(existingString) - byteLength(removedString)
+
+        let newLength = lengthOfRemainingExistingString + byteLength(replacementString)
+
+        if (newLength <= byteLimit) {
+            return true
+        }
+
+        // Don't allow any change if inserting a single char is already over the limit (typically this means typing)
+        if (replacementString.characters.count < 2) {
```

Comment:
Why special case this?

Reply:
Agreed that the behavior should be equivalent without this check, but this simple check for the common case (when a user is typing in their name) avoids the more complex logic/additional allocations required for the more complex case.